### PR TITLE
Causal

### PR DESCRIPTION
**Pull request by akashc1**

### CODE REVIEW

File: data.py [hunk 1]
```diff
@@ -27,7 +28,7 @@ def __init__(self, name, split, transforms, image_key, label_key='label'):
         except ImportError:
             raise ImportError('Please install datasets with `pip install datasets`.')
 
-        self.df = load_dataset(name, split=split)
+        self.df = load_dataset(name, split=split, cache_dir=DATA_DIR)
```
Comment (gmittal) [inline data.py]: would recommend using `export HF_DATASETS_CACHE` in `~/.bashrc` instead of manually setting it here.

https://huggingface.co/docs/datasets/cache#cache-directory

File: model.py [hunk 2]
```diff
@@ -215,17 +243,7 @@ def encode_mae(self, x, mask_ratio, rng):
...
-            self.class_embedding.astype(x.dtype)
-            + self.positional_embedding[:1, :].astype(x.dtype)
-        )
-        cls_token = jnp.broadcast_to(cls_token, (x.shape[0], 1, self.width))
-        x = jnp.concatenate([cls_token, x], axis=1)
-
-        x = self.ln_pre(x)
-        x = self.transformer(x)
-        x = self.ln_post(x)
+        x = self.fwd_transformer(x)
```
Comment (gmittal) [inline model.py:247]: This is just forwarding the encoder right?

File: model.py [hunk 3]
```diff
@@ -249,33 +267,54 @@ def decode_mae(self, x, ids_restore):
...
+        attn_mask = nn.make_causal_mask(jnp.ones((x.shape[0], T + 1), dtype=x.dtype))
+        x = self.causal_decoder(x, attn_mask)
+        x = self.dec_ln_post(x)
+        x = self.decoder_pred(x)
+        return x[:, 1:, :]
+
     def perturb(self, x, noise_std, rng):
-        noise = jax.random.normal(rng, shape=x.shape, dtype=x.dtype)
-        x = x + noise_std * noise
+        noise = noise_std * jax.random.normal(rng, shape=x.shape, dtype=x.dtype)
```
Comment (gmittal) [inline model.py:300]: `noise` is separated from `noise_std` as is described in this paper (should confirm that though):

https://arxiv.org/abs/2205.11423

File: data.py (hunk 1 above)
Comment (akashc1) [reply data.py]: fixed in `766e1054e6bf37335ed8c290a2324041bb33178c`

File: model.py (hunk 3 above)
Comment (akashc1) [reply model.py:300]: Yeah but in the loss we were supervising MSE between the standard normal noise & the scaled noise, so wasn't sure. Shouldn't really matter since it's just a single constant scale factor that'll be learned

File: model.py [hunk 4]
```diff
@@ -249,33 +267,54 @@ def decode_mae(self, x, ids_restore):
...
+        x = self.decoder_embed(x)
+        mask_tokens = jnp.tile(
+            self.mask_token,
+            (x.shape[0], T - x.shape[1] + 1, 1)
+        )
+        x = jnp.concatenate([x, mask_tokens], axis=1)
+
+        x = x + self.decoder_positional_embedding.astype(x.dtype)
+        x = self.dec_ln_pre(x)
+        attn_mask = nn.make_causal_mask(jnp.ones((x.shape[0], T + 1), dtype=x.dtype))
```
Comment (gmittal) [inline model.py:293]: Nice! So the encoder is non-causal and the decoder is causal, so it's basically a PrefixLM style thing. Could be really good for pretraining because the decoder sees way more data samples (parallelizes across the sequence for next-patch prediction) rather than 15% of the masked examples.

File: model.py (hunk 3 above)
Comment (gmittal) [reply model.py:300]: Hmmm... makes sense but I think we should investigate. It's weird enough that they choose to regress the noise rather than the pixel values. There are some ablations in that paper that we should revisit.

File: configs/ViT-B-16.py [hunk 5]
```diff
@@ -38,7 +38,7 @@ def get_config():
     config.image_key = 'image'
 
     # dataloader
-    config.num_workers = 48
+    config.num_workers = 8
```
Comment (gmittal) [inline configs/ViT-B-16.py:41]: why the reduction? are you seeing contention with the cores?

File: configs/pretrain_causal.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
+import ml_collections
+
+
+def get_config():
```
Comment (gmittal) [inline configs/pretrain_causal.py:4]: these are getting huge. you can subclass other configs by doing something like:

```python
from configs import pretrain_all  # or another base config

def get_config():
   config = pretrain_all.get_config()
   
   # change only what is different
   config.mae = False
```

Comment (gmittal) [reply configs/pretrain_causal.py:4]: We should probably do the same for finetuning configs, etc.

File: configs/ViT-B-16.py (hunk 5 above)
Comment (akashc1) [reply configs/ViT-B-16.py:41]: Yeah I kept getting some weird bus error, and this fixed it (but not sure 8 is the best value, just what worked)
